k2/csrc/ragged_test.cu

```diff
@@ -2028,4 +2029,74 @@ TEST(RaggedShapeOpsTest, GetPrefixesTest) {
   }
 }
 
+TEST(RaggedTest, AddSuffixToRaggedTest) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    {
+      // test with random large size
+      for (int32_t i = 0; i < 10; ++i) {
+        Ragged<int32_t> src = RandomRagged<int32_t>().To(context);
+        int32_t num_axes = src.NumAxes();
+        Array1<int32_t> suffix =
+            RandUniformArray1<int32_t>(context, src.TotSize(num_axes - 2),
+                                       0, 100);
+        Ragged<int32_t> dst = AddSuffixToRagged(src, suffix);
+        EXPECT_EQ(dst.NumAxes(), num_axes);
+        EXPECT_EQ(dst.NumElements(), src.NumElements() + suffix.Dim());
+        Ragged<int32_t> src_cpu = src.To(GetCpuContext());
+        Ragged<int32_t> dst_cpu = dst.To(GetCpuContext());
+        Array1<int32_t> suffix_cpu = suffix.To(GetCpuContext());
+        Array1<int32_t> dst_row_splits = dst_cpu.RowSplits(num_axes - 1);
+        for (RaggedShapeIndexIterator dst_iter = dst_cpu.shape.Iterator();
+            !dst_iter.Done(); dst_iter.Next()) {
+          const std::vector<int32_t> &dst_indexes = dst_iter.Value();
+          int32_t idx0 = dst_indexes[num_axes - 2];
+          int32_t size0 = dst_row_splits[idx0 + 1] - dst_row_splits[idx0];
+          ASSERT_GT(size0, 0);
+          if (dst_indexes.back() == size0 - 1) {
+            EXPECT_EQ(dst_cpu[dst_indexes], suffix_cpu[idx0]);
+          } else {
+            ASSERT_LT(dst_indexes.back(), size0 - 1);
+            EXPECT_EQ(dst_cpu[dst_indexes], src_cpu[dst_indexes]);
+          }
+        }
+      }
+    }
+  }
+}
+
+TEST(RaggedTest, AddPrefixToRaggedTest) {
+  for (auto &context : {GetCpuContext(), GetCudaContext()}) {
+    {
+      // test with random large size
+      for (int32_t i = 0; i < 10; ++i) {
+        Ragged<int32_t> src = RandomRagged<int32_t>().To(context);
+        int32_t num_axes = src.NumAxes();
+        Array1<int32_t> prefix =
+            RandUniformArray1<int32_t>(context, src.TotSize(num_axes - 2),
+                                       0, 100);
+        Ragged<int32_t> dst = AddPrefixToRagged(src, prefix);
+        EXPECT_EQ(dst.NumAxes(), num_axes);
+        EXPECT_EQ(dst.NumElements(), src.NumElements() + prefix.Dim());
+        Ragged<int32_t> src_cpu = src.To(GetCpuContext());
+        Ragged<int32_t> dst_cpu = dst.To(GetCpuContext());
+        Array1<int32_t> prefix_cpu = prefix.To(GetCpuContext());
+        Array1<int32_t> dst_row_splits = dst_cpu.RowSplits(num_axes - 1);
+        for (RaggedShapeIndexIterator dst_iter = dst_cpu.shape.Iterator();
```

Comment:
Ah, @freewym , I thought your previous testing approach (i.e. testing src first with indexes then testing prefix) is much more clear than this (as we don't need to write compact/very-high-performace code in testing, it's more important to make it clear.

Anyway, it's amazing that you use `RaggedShapeIndexIterator` to test this, I definitely like the testing approach with iterator, I just realized that I have not used it for a long time (the last time I used it is to test Transpose, Append, Stack stuff..)

Reply:
The test for prefix/suffix in the previous one is basically duplicate what we do in the functions; in this one I use the triplet to used to index all elements in dst including those from prefix/suffix, which I think is closer to the description of the functions

Reply:
(duplicate is fine as long as our testing depend on the interface instead of implementation). Anyway, this version is also fine with me,  it's up to you to change it or not. (and again, I like `Iterator` here). Will make this issue resloved